### ISSUE TITLE
RAFT: snapshot should use ByteBuffer rather than ByteArray

### ISSUE THREAD

**Issue by belaban** [belaban/jgroups-raft#174]
If possible, would be better to use direct `ByteBuffer` to hold the snapshot: having big `byte[]` is dangerous both for GC (causing humoungous allocations) and for the I/O, because native I/O don't understand byte[] heap objects and is forced to copy everything into a direct `ByteBuffer` (or worst, by using malloc to allocate one and free when completed)

**Comment by belaban** [comment]
@pruivo : can you check if `ByteBuffer` is used correctly/efficiently in `FileBasedLog`?

**Comment by pruivo** [comment]
https://github.com/belaban/jgroups-raft/commit/51e32faf94b58a30893fb491ebb34c56708eafeb

@belaban @franz1981 sorry again, but I don't see any advantage of using `ByteBuffer` over `ByteArray` in the commit above :(

**Comment by pruivo** [comment]
@belaban I didn't ask for anything :+1: Francesco did

**Comment by belaban** [comment]
oops :-)

**Comment by franz1981** [comment]
> sorry again, but I don't see any advantage of using ByteBuffer over ByteArray in the commit above :(

It's not the `ByteBuffer` vs `ByteArray` the problem, but just this part:

```java
os.write(snapshot.getArray(), snapshot.getOffset(), snapshot.getLength());
```
OutputStream are backed by posix I/O that is not able to understand what `byte[]` means, hence a copy is performed for each write. 
The copy is performed against a pooled direct `ByteBuffer` stored in a thread-local field; this can create severe stability issues and is not recommended (that's why i've implemented the file store reusing the same direct ByteBuffer over and over). 
In short, allocating a direct ByteBuffer is preferred and it's highly suggestefd to pool it somewhere, because releasing direct ByteBuffer is not as easy that with heap based ones (ie they use Cleaner/PhantomReference and can cause OOM of native memory with unpredictable results).

A note on using big byte[] in general:
The allocation of big byte[] in itself is already a problem depending by the length of arrays, because GC won't allow to allocate a single contiguos byte[] with easy, but trigger a so-called "humoungous allocation" that's way less friendly with GC.

**Comment by belaban** [comment]
So, are you happy with the change from `ByteArray` to `ByteBuffer`, @franz1981 ? :-)
The good thing is these are high-hanging fruit, as a snapshot is performed infrequently...

**Comment by franz1981** [comment]
>The good thing is these are high-hanging fruit, as a snapshot is performed infrequently...

I'm warning about the stability problem in itself, because being bitten by it several times before in my life - you can still use byte[] and leave the issue as it is, when/if the problem arise :+1:
